src/slim.cpp

```diff
@@ -0,0 +1,278 @@
+static BitcodeLib cpp_64_freebsd_aarch64("builtins-cpp-64-freebsd-aarch64.bc", TargetOS::freebsd, Arch::aarch64);
+#endif // defined(ISPC_FREEBSD_TARGET_ON) && defined(ISPC_ARM_ENABLED)
+
+#if defined(ISPC_FREEBSD_TARGET_ON) && defined(ISPC_X86_ENABLED)
+static BitcodeLib cpp_32_freebsd_i686("builtins-cpp-32-freebsd-i686.bc", TargetOS::freebsd, Arch::x86);
+static BitcodeLib cpp_64_freebsd_x86_64("builtins-cpp-64-freebsd-x86_64.bc", TargetOS::freebsd, Arch::x86_64);
+#endif // defined(ISPC_FREEBSD_TARGET_ON) && defined(ISPC_X86_ENABLED)
+
+#if defined(ISPC_WINDOWS_TARGET_ON) && defined(ISPC_ARM_ENABLED)
+static BitcodeLib cpp_64_windows_aarch64("builtins-cpp-64-windows-aarch64.bc", TargetOS::windows, Arch::aarch64);
+#endif // defined(ISPC_WINDOWS_TARGET_ON) && defined(ISPC_ARM_ENABLED)
+
+#if defined(ISPC_WINDOWS_TARGET_ON) && defined(ISPC_X86_ENABLED)
+static BitcodeLib cpp_32_windows_i686("builtins-cpp-32-windows-i686.bc", TargetOS::windows, Arch::x86);
+static BitcodeLib cpp_64_windows_x86_64("builtins-cpp-64-windows-x86_64.bc", TargetOS::windows, Arch::x86_64);
+#endif // defined(ISPC_WINDOWS_TARGET_ON) && defined(ISPC_X86_ENABLED)
+
+#if defined(ISPC_MACOS_TARGET_ON) && defined(ISPC_ARM_ENABLED)
+static BitcodeLib cpp_64_macos_arm64("builtins-cpp-64-macos-arm64.bc", TargetOS::macos, Arch::aarch64);
+#endif // defined(ISPC_MACOS_TARGET_ON) && defined(ISPC_ARM_ENABLED)
+
+#if defined(ISPC_MACOS_TARGET_ON) && defined(ISPC_X86_ENABLED)
+static BitcodeLib cpp_64_macos_x86_64("builtins-cpp-64-macos-x86_64.bc", TargetOS::macos, Arch::x86_64);
+#endif // defined(ISPC_MACOS_TARGET_ON) && defined(ISPC_X86_ENABLED)
+
+#if defined(ISPC_IOS_TARGET_ON) && defined(ISPC_ARM_ENABLED)
+// TODO! ??
+static BitcodeLib cpp_64_ios_arm64("builtins-cpp-64-ios-arm64.bc", TargetOS::ios, Arch::aarch64);
+#endif // defined(ISPC_IOS_TARGET_ON) && defined(ISPC_ARM_ENABLED)
+
+#if defined(ISPC_PS_TARGET_ON) && defined(ISPC_X86_ENABLED)
+static BitcodeLib cpp_64_ps4_x86_64("builtins-cpp-64-ps4-x86_64.bc", TargetOS::ps4, Arch::x86_64);
+#endif // defined(ISPC_PS_TARGET_ON) && defined(ISPC_X86_ENABLED)
+
+#if defined(ISPC_LINUX_TARGET_ON) && defined(ISPC_XE_ENABLED)
+static BitcodeLib cm_64_linux("builtins-cm-64.cpp", TargetOS::linux, Arch::xe64);
+#endif // defined(ISPC_LINUX_TARGET_ON) && defined(ISPC_XE_ENABLED)
+
+#if defined(ISPC_WINDOWS_TARGET_ON) && defined(ISPC_XE_ENABLED)
+static BitcodeLib cm_64_windows("builtins-cm-64.cpp", TargetOS::windows, Arch::xe64);
+#endif // defined(ISPC_WINDOWS_TARGET_ON) && defined(ISPC_XE_ENABLED)
+
+#ifdef ISPC_WASM_ENABLED
+static BitcodeLib cpp_32_web_wasm32("builtins-cpp-32-web-wasm32.cpp", TargetOS::web, Arch::wasm32);
+static BitcodeLib cpp_64_web_wasm64("builtins-cpp-64-web-wasm64.cpp", TargetOS::web, Arch::wasm64);
+#endif // ISPC_WASM_ENABLED
+
+// Target-specific built-in bitcode_libs
+// clang-format off
+#if defined(ISPC_UNIX_TARGET_ON) && defined(ISPC_X86_ENABLED)
```

Comment:
Is it possible to simplify the patterns below with, maybe, macro magic?

Reply:
Under assumption that target name in file contains `_` instead of `-`, it is possible to _simplify_ with something like this (not really tested though):
```c
#define SSE2_TARGETS(M) \
  M(sse2_i32x4); \
  M(sse2_i32x8);

#define SSE4_TARGETS(M) \
  M(sse4_i8x16) \
  M(sse4_i16x8) \
  M(sse4-i32x4) \
  M(sse4-i32x8)

#define AVX1_TARGETS(M) \
  M(avx1-i32x8) \
  M(avx1-i32x16) \
  M(avx1-i64x4)

#define AVX2_TARGETS(M) \
  M(avx2-i8x32) \
  M(avx2-i16x16) \
  M(avx2-i32x4) \
  M(avx2-i32x8) \
  M(avx2-i32x16) \
  M(avx2-i64x4)

#define AVX512KNL_TARGETS(M) \
  M(avx512knl-x16)

#define AVX512SKX_TARGETS(M) \
  M(avx512skx-x4) \
  M(avx512skx-x8) \
  M(avx512skx-x16) \
  M(avx512skx-x32) \
  M(avx512skx-x64)

#define AVX512SPR_TARGETS(M) \
  M(avx512spr-x4) \
  M(avx512spr-x8) \
  M(avx512spr-x16) \
  M(avx512spr-x32) \
  M(avx512spr-x64)

#define X86_TARGETS(M) \
  SSE2_TARGETS(M) \
  SSE4_TARGETS(M) \
  AVX1_TARGETS(M) \
  AVX2_TARGETS(M) \
  AVX512KNL_TARGETS(M) \
  AVX512SKX_TARGETS(M) \
  AVX512SPR_TARGETS(M)

#define TARGET_BITCODE_LIB(TARGET, OS_NAME, OS_TARGET, BIT, ARCH) static BitcodeLib target_ ## TARGET ## _ ## BIT ## bit_ ## OS_NAME("builtins-" #TARGET "-" #BIT "bit-" #OS_NAME ".bc", ISPCTarget::TARGET, TargetOS::OS_TARGET, Arch::ARCH)

#define TARGET_BITCODE_LIB_UNIX_X86_64(TARGET) TARGET_BITCODE_LIB(TARGET, unix, linux, 64, x86_64)
#define TARGET_BITCODE_LIB_UNIX_X86(TARGET) TARGET_BITCODE_LIB(TARGET, unix, linux, 32, x86)

#define TARGET_BITCODE_LIB_WINDOWS_X86_64(TARGET) TARGET_BITCODE_LIB(TARGET, windows, windows, 64, x86_64)
#define TARGET_BITCODE_LIB_WINDOWS_X86(TARGET) TARGET_BITCODE_LIB(TARGET, windows, windows, 32, x86)

X86_TARGETS(TARGET_BITCODE_LIB_UNIX_X86_64)
X86_TARGETS(TARGET_BITCODE_LIB_UNIX_X86)

X86_TARGETS(TARGET_BITCODE_LIB_WINDOWS_X86_64)
X86_TARGETS(TARGET_BITCODE_LIB_WINDOWS_X86)
```

Reply:
It might be worth considering if the future extensions or support of this code would be easier.

Reply:
Maybe we can use CMake configuration template here?

Reply:
> Maybe we can use CMake configuration template here?

What for?

Reply:
> What for?

To avoid manually maintaining the list of builtins, which is prone to errors. I think it can be more efficient to utilize CMake. CMake is already aware of all target names, architectures, and compilation definitions. It would also help to concentrate all logic regarding build in one place.

Reply:
> To avoid manually maintaining the list of builtins, which is prone to errors.
I doubt that CMake code would be free of it. 

> I think it can be more efficient to utilize CMake. CMake is already aware of all target names, architectures, and compilation definitions. It would also help to concentrate all logic regarding build in one place.

I disagree that this is build logic.

Reply:
I still can hardly imagine how to support the list of `BitcodeLib` in binary_slim.cpp up-to-date and in sync with logic in GenerateBuiltins.cmake `generate_common_builtins` function.
I'd do something like this:
```
# GenerateBuiltins.cmake
file(WRITE ${CMAKE_BINARY_DIR}/bitcode_libs_generated.cpp "

macro(write_bitcode_lib bitcode_file)
    file(APPEND ${CMAKE_BINARY_DIR}/bitcode_libs_generated.cpp
        "static BitcodeLib ${bitcode_file};\n")
endmacro()

function(builtin_to_cpp bit os generic_arch)
...
write_bitcode_lib(${bc})
```
Same for targets.